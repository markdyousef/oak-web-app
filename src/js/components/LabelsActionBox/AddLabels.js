// @flow
import React from 'react';
import styled from 'styled-components';
import colors from '../../styles/colors';
import Pencil from '../../icons/pencil';
import { Box as Label } from '../shared/Label';

const SectionTitle = styled.div`
    font-size: 16px;
    font-weight: bold;
    padding-bottom: 8px;
`;

const Section = styled.div`
    font-size: 15px;
`;

const LabelsSection = styled(Section)`
    padding: 8px 0 16px;
    width: calc(100% + 20px);
    overflow-y: auto;
    max-height: 300px;
    margin-left: -20px;
`;

const NoLabels = styled.div`
    font-size: 14px;
    text-align: center;
    padding: 16px 0 24px;
    font-weight: normal;
    line-height: 1.4;
    color: rgba(19, 21, 23, 0.5);
`;

const LabelContainer = styled.div`
    width: 100%;
    display: flex;
    margin-top: -20px;
    position: relative;
`;

const ActionSection = styled.section`
    border-top: 1px solid ${colors.lightGrey};
    padding-top: 16px;
    display: flex;
    justify-content: flex-end;
    margin-right: -20px;
    margin-left: -20px;
    padding-right: 20px;
    font-weight: normal;
`;

const Button = styled.button`
    background: transparent;
    padding: 0;
    border: none;
    font-size: 14px;
    margin: 15px 0 !important;
    cursor: pointer;
`;

const Edit = styled.div`
    position: absolute;
    top: 22px;
    right: 6px;
    z-index: 1;
`;

type Props = {
    cardLabels: ?Array<string>,
    collectionLabels: ?Array<Object>,
    changePage: Function,
    onSelect: Function
}

export default ({ cardLabels, collectionLabels, changePage, onSelect, showEdit }: Props) => {
    return (
        <div>
            <SectionTitle>
                Add label:
            </SectionTitle>
            <Section>
                {(collectionLabels && collectionLabels.length > 0) ?
                    <LabelsSection>
                        {collectionLabels.map((label) => {
                            const isActive =
                                cardLabels &&
                                cardLabels.findIndex(id => id === label.id) > -1;
                            return (
                                <LabelContainer
                                    key={label.id + label.name}
                                >
                                    <Label
                                        onClick={() => onSelect(label.id)}
                                        style={{ backgroundColor: label.color }}
                                        name={label.name}
                                        isActive={isActive}
                                    />
                                    <Edit onClick={() => showEdit(label)}>
                                        <Pencil />
                                    </Edit>
                                </LabelContainer>
                            );
                        })}
                    </LabelsSection>
                    :
                    <NoLabels>Your team has not created any labels in this collection. Create your first label below.</NoLabels>
                }
            </Section>
            <ActionSection>
                <div onClick={changePage}>
                    Create a new label
                </div>
            </ActionSection>
        </div>
    );
};
