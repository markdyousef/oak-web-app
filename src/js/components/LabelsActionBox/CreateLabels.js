import React, { Component, PropTypes } from 'react';
import Input from '../shared/Input';

import css from './LabelsActionBox.css';

class CreateLabels extends Component {
    constructor() {
        super();
        this.state = {
            name: ''
        };
    }
    render() {
        const { name } = this.state;
        return (
            <div className={css.createContainer}>
                <header>
                    <h3>Create Label</h3>
                </header>
                <div className={css.name}>
                    <Input
                        title="NAME"
                        value={name}
                        onChange={value => this.setState({ name: value })}
                    />
                </div>
            </div>
        );
    }
}

export default CreateLabels;
