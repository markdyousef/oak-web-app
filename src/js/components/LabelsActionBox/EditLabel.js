// @flow
import React from 'react';
import styled from 'styled-components';
import Input from '../shared/Input';
import { SquareButton } from '../shared/Button';
import colors from '../../styles/colors';
import BackIcon from '../../icons/back';
import CheckIcon from '../../icons/checkmark';

const Grid = styled.div`
    display: flex;
    flex-wrap: wrap;
    width: calc(100% + 10px);
    flex-grow: 1;
    margin: 0px 0 0 -10px;
`;

const Label = styled.div`
    width: 30px;
    height: 33.5px;
    border-radius: 3px;
    cursor: pointer;
    margin: 5px;
    display: flex;
    align-items: center;
    justify-content: center;
    width: calc(100% * (1/5) - 10px - 1px);
    flex-grow: 1;
    margin: 10px 0 0 10px;
    & svg {
        fill: ${colors.white};
        position: relative;
        top: 1px;
    }
`;

const Section = styled.section`
    margin: 16px 0;
    & h5 {
        font-size: 12px;
        font-weight: lighter;
        color: ${colors.grey};
        font-family: 'proxima-nova'
    }
`;

const ActionSection = styled.section`
    border-top: 1px solid ${colors.lightGrey};
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 16px 20px 0px;
    margin-left: -20px;
    margin-right: -20px;
`;

const Back = styled.button`
    cursor: pointer;
    border: none;
    background-color: #fff;
    padding: 0;
    font-weight: normal;
    font-size: 13px;
    display: flex;
    align-items: center;
    & svg {
        margin-right: 6px;
        height: 13px;
        margin-left: -2px;
    }
`;

const Save = styled.button`
    cursor: pointer;
    border: none;
    background-color: #fff;
    padding: 0;
    font-size: 14px;
`;

const Delete = styled(Save)`
    color: #131517;
`;

type Props = {
    onCreate?: Function,
    onChange: Function,
    changePage: Function,
    labelName: ?string,
    selectedColor: string,
    labelColors: Array<string>
}

export default ({ ...props }:Props) => {
    const {
        changePage,
        labelName,
        onChange,
        labelColors,
        selectedColor,
        onCreate,
        onUpdate,
        onDelete
    } = props;
    return (
        <div>
            <Back onClick={changePage}>
                <BackIcon />
                Back
            </Back>
            <Section>
                <Input
                    title="NAME"
                    value={labelName || ''}
                    onChange={name => onChange({ name })}
                    placeholder="Label name"
                />
            </Section>
            <Section>
                <h5>SELECT COLOR: </h5>
                <Grid>
                    {labelColors.map((color) => {
                        const isSelected = (selectedColor === color);
                        return (
                            <Label
                                style={{ backgroundColor: color }}
                                key={color}
                                onClick={() => onChange({ color })}
                            >
                                {isSelected && <CheckIcon />}
                            </Label>
                        );
                    }
                    )}
                </Grid>
            </Section>
            <ActionSection>
                {onCreate &&
                    <Save onClick={onCreate}>
                        Save label
                    </Save>
                }
                {onDelete &&
                    <Delete onClick={onDelete}>
                        Delete
                    </Delete>
                }
                {onUpdate &&
                    <Save onClick={onUpdate}>
                        Savel label
                    </Save>
                }
            </ActionSection>
        </div>
    );
};
