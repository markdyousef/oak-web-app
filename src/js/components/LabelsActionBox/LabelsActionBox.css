.container {
    /*position: absolute;*/
    width: 250px;
    background-color: #fff;
    z-index: 2000;
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
}

.container header {
    border-bottom: 1px solid #ddd;
    height: 30px;
    margin: 10px;
    display: flex;
    align-items: center;
    padding: 10px;
    width: 100%;
}

.container header h3 {
    font-size: 12px;
}

/* CREATE LABELS */

.createContainer {
    width: 100%;
    height: 400px;
    display: flex;
    flex-direction: column;
    align-items: center;
}

.name {
    margin-top: 10px;
    width: 80%;
}
