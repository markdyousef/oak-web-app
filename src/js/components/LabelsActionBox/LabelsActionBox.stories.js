import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import LabelsActionBox from './LabelsActionBox';

storiesOf('LabelsActionBox', module)
    .add('LabelsActionBox', () => (
        <LabelsActionBox
            close={action('close')}
            create={action('create')}
        />
    ));
