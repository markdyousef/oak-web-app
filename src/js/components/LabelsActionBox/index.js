import LabelsActionBox from './LabelsActionBox';

export default LabelsActionBox;
