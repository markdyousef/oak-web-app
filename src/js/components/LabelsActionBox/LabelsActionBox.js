// @flow
import React, { Component } from 'react';
import styled from 'styled-components';
import CreateLabels from './CreateLabels';
import AddLabels from './AddLabels';
import Dropdown from '../shared/Dropdown';
import { labelColors } from '../../styles';

const Container = styled.div`
    position: absolute;
    z-index: 9999;
    margin-top: 20px;
    width: 250px;
`;

type Data = {
    loading: bool,
    refetch: Function,
    grove?: {
        labels: Array<Object>
    }
}

type Props = {
    collectionId: string,
    changeCardLabel: Function,
    createLabel: Function,
    collection?: Data,
    card?: Object,
    labels?: Object,
    updateLabels: (field:Object) => void
};

type State = {};
type DefaultProps = {};

class LabelsActionBox extends Component<DefaultProps, Props, State> {
    static defaultProps: DefaultProps;
    state: State;
    componentWillReceiveProps(nextProps:Props) {
        const { collection, card, labels } = nextProps;
        const didInitialize = labels && labels.get('didInitialize');

        // TODO: find a better way to load into redux
        if (!collection || collection.loading) return;
        if (!didInitialize) {
            this.updateLabel('didInitialize', true);
            // add collectionLabels to redux
            if (collection.grove) {
                const { grove } = collection;
                this.updateLabel(
                    'collectionLabels',
                    (grove.labels) || []
                );
            }
            // add cardLabels to redux
            if (!card || card.loading) return;
            if (card.seed) {
                const { seed } = card;
                this.updateLabel(
                    'cardLabels',
                    (seed.labels) || []
                )
            }
        }
    }
    createLabel = (labelName:string, labelColor:string) => {
        const { createLabel, collectionId, create } = this.props;
        createLabel(collectionId, labelName, labelColor)
            .then((res) => {
                const { name, color, id } = res.data.createLabel;
                create({ name, color, id });
            })
            .catch(() => {
                const message = {
                    type: 'error',
                    message: "We couldn't create your label",
                    onSave: () => this.createLabel(labelName, labelColor)
                };
                this.updateLabel('message', message);
            });
    }
    updateLabel = (key:string, value: any) => {
        const { updateLabels } = this.props;
        updateLabels({ key, value });
    }
    onClose = () => this.updateLabel('showLabels', false);
    changePage = (page: string) => this.updateLabel('page', page);
    changeCardLabel = (labelId: string) => {
        const {
            createCard,
            removeLabel,
            attach,
            detach,
            addLabel,
            updateCard,
            cardId,
            collectionId,
            labels
        } = this.props;
        if (!cardId) {
            createCard(collectionId, '')
                .then((res) => {
                    const id = res.data.createSeed.id;
                    updateCard({
                        key: 'cardId',
                        value: id
                    });
                    this.changeCardLabel(labelId);
                })
                .catch(() => {
                    const message = {
                        type: 'error',
                        message: "We couldn't create your label",
                        onSave: () => this.changeCardLabel(labelId)
                    };
                    this.updateLabel('message', message);
                });
            return;
        }
        const cardLabels = labels.get('cardLabels');
        const labelExist = cardLabels.findIndex(id => id === labelId) > -1;
        if (labelExist) {
            removeLabel(cardId, labelId)
                .then((res) => {
                    if (res.data.removeSeedLabel) {
                        detach(labelId);
                    }
                    updateCard({
                        key: 'shouldUpdate',
                        value: true
                    });
                })
                .catch(() => {
                    const message = {
                        type: 'error',
                        message: "We couldn't remove your label",
                        onSave: () => this.changeCardLabel(labelId)
                    };
                    this.updateLabel('message', message);
                });
        } else {
            addLabel(cardId, labelId)
                .then((res) => {
                    if (res.data.addSeedLabel) {
                        attach(labelId);
                    }
                    updateCard({
                        key: 'shouldUpdate',
                        value: true
                    });
                })
                .catch(() => {
                    const message = {
                        type: 'error',
                        message: "We couldn't add your label",
                        onSave: () => this.changeCardLabel(labelId)
                    };
                    this.updateLabel('message', message);
                });
        }
    }
    renderLabels = () => {
        const { labels } = this.props;
        const labelsObj = (labels) ? labels.toJS() : {};

        switch (labelsObj.page) {
        case 'CREATE':
            return (
                <CreateLabels
                    onCreate={this.createLabel}
                    changePage={() => this.changePage('ADD')}
                    onChange={this.updateLabel}
                    labelName={labelsObj.labelName}
                    selectedColor={labelsObj.selectedColor}
                    labelColors={labelColors}
                />
            );
        case 'ADD':
            return (
                <AddLabels
                    changePage={() => this.changePage('CREATE')}
                    collectionLabels={labelsObj.collectionLabels}
                    cardLabels={labelsObj.cardLabels}
                    onSelect={this.changeCardLabel}
                />
            );
        case 'EDIT':
            return null;
        default:
            return null;
        }
    }
    render() {
        return (
            <Container>
                <Dropdown arrowPos="left" onClose={this.onClose}>
                    {this.renderLabels()}
                </Dropdown>
            </Container>
        );
    }
}

export default LabelsActionBox;
