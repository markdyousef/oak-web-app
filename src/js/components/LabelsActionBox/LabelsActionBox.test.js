// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import LabelsActionBox from './LabelsActionBox';

it('renders correctly', () => {
    const props = {
        changeCardLabel: () => {},
        collectionId: '1',
        createLabel: () => {},
        updateLabels: () => {},
        data: {
            loading: true,
            refetch: () => {}
        }

    };
    const component = renderer.create(
        <LabelsActionBox {...props} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
