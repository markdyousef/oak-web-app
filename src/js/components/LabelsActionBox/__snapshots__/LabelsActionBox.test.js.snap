exports[`test renders correctly 1`] = `
<div
  className="gTWMPT">
  <div>
    <div
      className="hqCZJo"
      style={
        Object {
          "left": "25px",
        }
      } />
    <div
      className="fEcnKi" />
  </div>
</div>
`;
